request_headers.go

```diff
@@ -0,0 +1,115 @@
+package abstractions
+
+import "strings"
+
+//RequestHeaders represents a collection of request headers
+type RequestHeaders struct {
+	headers map[string]map[string]bool
+}
```

Comment:
In this case can we consider using an empty struct instead of a Boolean to be memory conservative in the "set" implementation i.e 

```
//RequestHeaders represents a collection of request headers
type RequestHeaders struct {
	headers map[string]map[string]struct{}
}
type void struct{}
var Void void
```

and within functions 

```
func (r *RequestHeaders) Add(key string, value string, additionalValues ...string) {
	if key == "" || value == "" {
		return
	}
	if r.headers == nil {
		r.headers = make(map[string]map[string]struct{})
	}
	normalizedKey := r.normalizeHeaderKey(key)
	if r.headers[normalizedKey] == nil {
		r.headers[normalizedKey] = make(map[string]struct{})
	}
	r.headers[normalizedKey][value] = Void
	for _, v := range additionalValues {
		r.headers[normalizedKey][v] = Void
	}
}

```